calyx-opt/src/passes/external_to_ref.rs

```diff
@@ -0,0 +1,53 @@
+use crate::traversal::{Action, ConstructVisitor, Named, VisResult, Visitor};
+use calyx_ir::{self as ir, GetAttributes, LibrarySignatures};
+use calyx_utils::CalyxResult;
+
+//Turns memory cell primitives with the `@external(1)` attribute into
+// `ref` memory cells without the `@external` attribute.
+pub struct ExternalToRef;
+
+impl Named for ExternalToRef {
+    fn name() -> &'static str {
+        "external-to-ref"
+    }
+
+    fn description() -> &'static str {
+        "Turn memory cells marked with `@external(1) into `ref` memory cells."
+    }
+}
+
+//TODO: Is this right? Not sure this is what we want
+impl ConstructVisitor for ExternalToRef {
+    fn from(_ctx: &ir::Context) -> CalyxResult<Self>
+    where
+        Self: Sized,
+    {
+        let external_to_ref = ExternalToRef;
+        Ok(external_to_ref)
+    }
+
+    fn clear_data(&mut self) {
+        //TODO: do we need this?
+    }
+}
```

Comment:
Confession: This was entirely pattern matched from other passes (I think `externalize`) and I do not understand what it is doing. Happy to hear about any changes that should be made.

Reply:
I'm afraid I don't have the answer to this one—perhaps somebody else knows whether this is required. It sure seem like it is required for every pass to me; I would feel comfortable assuming that this is exactly what you need.

Reply:
cc @rachitnigam When you have a moment, can you confirm if this looks right?